// @recogito/text-annotator/dist/text-annotator.es.js
import { colord as J } from "colord";
import { dequal as at } from "dequal/lite";
import { v4 as tt } from "uuid";
import { serializeW3CBodies as ct, parseW3CBodies as lt, parseW3CUser as dt, createStore as ut, createSelectionState as ft, createHoverState as gt, createViewportState as mt, Origin as K, createAnonymousGuest as pt, createUndoStack as ht, createLifecycleObserver as yt, createBaseAnnotator as bt } from "@annotorious/core";
import { Origin as be, UserSelectAction as Se, createBody as Ce } from "@annotorious/core";
import St from "rbush";
import D from "hotkeys-js";
import { poll as Ct } from "poll";
const et = "not-annotatable", U = `.${et}`, F = (t, n) => {
  var o;
  return t.contains(n) ? !!(n instanceof HTMLElement ? n.closest(U) : (o = n.parentElement) == null ? void 0 : o.closest(U)) : !0;
}, wt = (t, n) => {
  const e = n.commonAncestorContainer;
  return !F(t, e);
}, Et = (t) => t.addEventListener("click", (n) => {
  // Allow clicks within not-annotatable elements
  !n.target.closest(U) && !n.target.closest("a") && n.preventDefault();
}), z = (t) => ({
  ...t,
  type: t.type,
  x: t.x,
  y: t.y,
  clientX: t.clientX,
  clientY: t.clientY,
  offsetX: t.offsetX,
  offsetY: t.offsetY,
  screenX: t.screenX,
  screenY: t.screenY,
  isPrimary: t.isPrimary,
  altKey: t.altKey,
  ctrlKey: t.ctrlKey,
  metaKey: t.metaKey,
  shiftKey: t.shiftKey,
  button: t.button,
  buttons: t.buttons,
  currentTarget: t.currentTarget,
  target: t.target,
  defaultPrevented: t.defaultPrevented,
  detail: t.detail,
  eventPhase: t.eventPhase,
  pointerId: t.pointerId,
  pointerType: t.pointerType,
  timeStamp: t.timeStamp
}), V = (t) => ({
  ...t,
  type: t.type,
  key: t.key,
  code: t.code,
  location: t.location,
  repeat: t.repeat,
  altKey: t.altKey,
  ctrlKey: t.ctrlKey,
  metaKey: t.metaKey,
  shiftKey: t.shiftKey,
  currentTarget: t.currentTarget,
  target: t.target,
  defaultPrevented: t.defaultPrevented,
  detail: t.detail,
  timeStamp: t.timeStamp
}), At = typeof navigator < "u" && // @ts-ignore
/mac/i.test(navigator.userAgentData ? navigator.userAgentData.platform : navigator.platform), xt = (t) => {
  !t.hasAttribute("tabindex") && t.tabIndex < 0 && t.setAttribute("tabindex", "-1"), t.classList.add("no-focus-outline");
}, W = (t, n = 10) => {
  let e;
  return ((...o) => {
    clearTimeout(e), e = setTimeout(() => t.apply(void 0, o), n);
  });
}, vt = function* (t) {
  const n = document.createNodeIterator(
    t.commonAncestorContainer,
    NodeFilter.SHOW_ELEMENT,
    (o) => o instanceof HTMLElement && o.classList.contains(et) && !o.parentElement.closest(U) && t.intersectsNode(o) ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_SKIP
  );
  let e;
  for (; e = n.nextNode(); )
    e instanceof HTMLElement && (yield e);
}, Tt = (t, n) => {
  if (!wt(t, n)) return [];
  const e = [];
  let o = null;
  for (const s of vt(n)) {
    let r;
    o ? (r = document.createRange(), r.setStartAfter(o), r.setEndBefore(s)) : (r = n.cloneRange(), r.setEndBefore(s)), r.collapsed || e.push(r), o = s;
  }
  if (o) {
    const s = n.cloneRange();
    s.setStartAfter(o), s.collapsed || e.push(s);
  }
  return e.length > 0 ? e : [n];
}, j = (t) => {
  const n = t.cloneContents();
  return n.querySelectorAll(U).forEach((e) => e.remove()), n;
}, Rt = (t, n, e = 10, o) => {
  const s = o ? t.startContainer.parentElement.closest(o) : n, r = document.createRange();
  r.setStart(s, 0), r.setEnd(t.startContainer, t.startOffset);
  const l = j(r).textContent, u = document.createRange();
  u.setStart(t.endContainer, t.endOffset), s === document.body ? u.setEnd(s, s.childNodes.length) : u.setEndAfter(s);
  const g = j(u).textContent;
  return {
    prefix: l.substring(l.length - e),
    suffix: g.substring(0, e)
  };
}, P = (t) => t.every((n) => n.range instanceof Range && !n.range.collapsed), Lt = /^\s*$/, kt = (t) => Lt.test(t.toString()), Ot = (t, n) => {
  const e = (r) => Math.round(r * 10) / 10, o = {
    top: e(t.top),
    bottom: e(t.bottom),
    left: e(t.left),
    right: e(t.right)
  }, s = {
    top: e(n.top),
    bottom: e(n.bottom),
    left: e(n.left),
    right: e(n.right)
  };
  if (Math.abs(o.top - s.top) < 0.5 && Math.abs(o.bottom - s.bottom) < 0.5) {
    if (Math.abs(o.left - s.right) < 0.5 || Math.abs(o.right - s.left) < 0.5)
      return "inline-adjacent";
    if (o.left >= s.left && o.right <= s.right)
      return "inline-is-contained";
    if (o.left <= s.left && o.right >= s.right)
      return "inline-contains";
  } else if (o.top <= s.top && o.bottom >= s.bottom) {
    if (o.left <= s.left && o.right >= s.right)
      return "block-contains";
  } else if (o.top >= s.top && o.bottom <= s.bottom && o.left >= s.left && o.right <= s.right)
    return "block-is-contained";
}, It = (t, n) => {
  const e = Math.min(t.left, n.left), o = Math.max(t.right, n.right), s = Math.min(t.top, n.top), r = Math.max(t.bottom, n.bottom);
  return new DOMRect(e, s, o - e, r - s);
}, Bt = (t) => t.reduce((n, e) => {
  if (e.width === 0 || e.height === 0)
    return n;
  let o = [...n], s = !1;
  for (const r of n) {
    const l = Ot(e, r);
    if (l === "inline-adjacent") {
      o = o.map((u) => u === r ? It(e, r) : u), s = !0;
      break;
    } else if (l === "inline-contains") {
      o = o.map((u) => u === r ? e : u), s = !0;
      break;
    } else if (l === "inline-is-contained") {
      s = !0;
      break;
    } else if (l === "block-contains" || l === "block-is-contained") {
      e.width < r.width && (o = o.map((u) => u === r ? e : u)), s = !0;
      break;
    }
  }
  return s ? o : [...o, e];
}, []), ge = (t) => ({
  length: t.length,
  item: (n) => t[n],
  [Symbol.iterator]: function* () {
    for (let n = 0; n < this.length; n++)
      yield this.item(n);
  }
}), Nt = (t, n, e) => {
  const o = document.createRange(), s = e ? t.startContainer.parentElement.closest(e) : n;
  o.setStart(s, 0), o.setEnd(t.startContainer, t.startOffset);
  const r = j(o).textContent, l = t.toString(), u = r.length || 0, g = u + l.length;
  return e ? { quote: l, start: u, end: g, range: t, offsetReference: s } : { quote: l, start: u, end: g, range: t };
}, nt = (t, n) => {
  var p, d;
  const { start: e, end: o } = t, s = t.offsetReference || n, r = document.createNodeIterator(
    n,
    NodeFilter.SHOW_TEXT,
    (i) => {
      var E;
      return (E = i.parentElement) != null && E.closest(U) ? NodeFilter.FILTER_SKIP : NodeFilter.FILTER_ACCEPT;
    }
  );
  let l = 0;
  const u = document.createRange();
  let g = r.nextNode();
  g === null && console.error("Could not revive annotation target. Content missing.");
  let S = !s;
  for (; g !== null; ) {
    if (S || (S = typeof (s == null ? void 0 : s.contains) == "function" ? s.contains(g) : !1), S) {
      const i = ((p = g.textContent) == null ? void 0 : p.length) || 0;
      if (l + i > e) {
        u.setStart(g, e - l);
        break;
      }
      l += i;
    }
    g = r.nextNode();
  }
  for (; g !== null; ) {
    const i = ((d = g.textContent) == null ? void 0 : d.length) || 0;
    if (l + i >= o) {
      u.setEnd(g, o - l);
      break;
    }
    l += i, g = r.nextNode();
  }
  return {
    ...t,
    range: u
  };
}, Y = (t, n) => P(t.selector) ? t : {
  ...t,
  selector: t.selector.map((e) => e.range instanceof Range && !e.range.collapsed ? e : nt(e, n))
}, q = (t, n) => P(t.target.selector) ? t : { ...t, target: Y(t.target, n) }, Pt = (t, n) => {
  const e = t.cloneRange();
  return n.contains(e.startContainer) || e.setStart(n, 0), n.contains(e.endContainer) || e.setEnd(n, n.childNodes.length), e;
}, ot = (t) => {
  if (t === null)
    return document.scrollingElement;
  const { overflowY: n } = window.getComputedStyle(t);
  return n !== "visible" && n !== "hidden" && t.scrollHeight > t.clientHeight ? t : ot(t.parentElement);
}, Mt = (t, n) => (e) => {
  const o = typeof e == "string" ? e : e.id, s = (p) => {
    const d = r.getBoundingClientRect(), i = r.clientHeight, E = r.clientWidth, C = p.selector[0].range.getBoundingClientRect(), { width: v, height: x } = n.getAnnotationBounds(o), a = C.top - d.top, c = C.left - d.left, b = r.parentElement ? r.scrollTop : 0, y = r.parentElement ? r.scrollLeft : 0, m = a + b - (i - x) / 2, f = c + y - (E - v) / 2;
    r.scroll({ top: m, left: f, behavior: "smooth" });
  }, r = ot(t);
  if (!r)
    return console.warn(`The scroll parent is missing for the annotation: ${o}`, { container: t }), !1;
  const l = n.getAnnotation(o);
  if (!l)
    return console.warn(`The annotation is missing in the store: ${o}`), !1;
  const { range: u } = l.target.selector[0];
  if (u && !u.collapsed)
    return s(l.target), !0;
  const g = Y(l.target, t), { range: S } = g.selector[0];
  return S && !S.collapsed ? (s(g), !0) : !1;
}, M = {
  fill: "rgb(0, 128, 255)",
  fillOpacity: 0.18
}, X = {
  fill: "rgb(0, 128, 255)",
  fillOpacity: 0.45
}, Ht = (t, n, e, o, s) => {
  var l, u;
  const r = e ? typeof e == "function" ? e(t.annotation, t.state, s) || ((l = t.state) != null && l.selected ? X : M) : e : (u = t.state) != null && u.selected ? X : M;
  return o && o.paint(t, n) || r;
}, Kt = (t) => {
  const { top: n, left: e } = t.getBoundingClientRect(), { innerWidth: o, innerHeight: s } = window, r = -e, l = -n, u = o - e, g = s - n;
  return { top: n, left: e, minX: r, minY: l, maxX: u, maxY: g };
}, Ut = (t) => {
  let n = /* @__PURE__ */ new Set();
  return (o) => {
    const s = o.map((r) => r.id);
    (n.size !== s.length || s.some((r) => !n.has(r))) && t.set(s), n = new Set(s);
  };
}, G = (t, n, e, o) => {
  const { store: s, selection: r, hover: l } = n;
  let u, g, S;
  const p = Ut(e), d = (R) => {
    const { x: h, y: A } = t.getBoundingClientRect(), L = s.getAt(R.clientX - h, R.clientY - A, !1, g);
    L ? l.current !== L.id && (t.classList.add("hovered"), l.set(L.id)) : l.current && (t.classList.remove("hovered"), l.set(null));
  };
  t.addEventListener("pointermove", d);
  const i = (R = !1) => {
    S && S.clear();
    const h = Kt(t), { minX: A, minY: L, maxX: B, maxY: N } = h, I = g ? s.getIntersecting(A, L, B, N).filter(({ annotation: O }) => g(O)) : s.getIntersecting(A, L, B, N), _ = r.selected.map(({ id: O }) => O), k = I.map(({ annotation: O, rects: $ }) => {
      const H = _.includes(O.id), it = O.id === l.current;
      return {
        annotation: O,
        rects: $,
        state: { selected: H, hovered: it }
      };
    });
    o.redraw(k, h, u, S, R), setTimeout(() => p(I.map(({ annotation: O }) => O)), 1);
  }, E = (R) => {
    S = R, i();
  }, C = (R) => {
    u = R, i();
  }, v = (R) => {
    g = R, i(!1);
  }, x = () => i();
  s.observe(x);
  const a = r.subscribe(() => i()), c = l.subscribe(() => i()), b = () => i(!0);
  document.addEventListener("scroll", b, { capture: !0, passive: !0 });
  const y = W(() => {
    s.recalculatePositions(), S && S.reset(), i();
  });
  window.addEventListener("resize", y);
  const m = new ResizeObserver(y);
  m.observe(t);
  const f = { attributes: !0, childList: !0, subtree: !0 }, w = new MutationObserver(W((R) => {
    R.every((A) => A.target === t || t.contains(A.target)) || i(!0);
  }, 150));
  return w.observe(document.body, f), {
    destroy: () => {
      t.removeEventListener("pointermove", d), o.destroy(), s.unobserve(x), a(), c(), document.removeEventListener("scroll", b), window.removeEventListener("resize", y), m.disconnect(), w.disconnect();
    },
    redraw: i,
    setStyle: C,
    setFilter: v,
    setPainter: E,
    setVisible: o.setVisible
  };
}, $t = () => {
  const t = document.createElement("canvas");
  return t.width = window.innerWidth, t.height = window.innerHeight, t.className = "r6o-canvas-highlight-layer bg", t;
}, Ft = (t, n) => {
  t.width = window.innerWidth, t.height = window.innerHeight;
}, _t = (t) => {
  t.classList.add("r6o-annotatable");
  const n = $t(), e = n.getContext("2d");
  document.body.appendChild(n);
  const o = (u, g, S, p) => requestAnimationFrame(() => {
    const { width: d, height: i } = n;
    e.clearRect(-0.5, -0.5, d + 1, i + 1), p && p.clear();
    const { top: E, left: C } = g;
    [...u].sort((x, a) => {
      const { annotation: { target: { created: c } } } = x, { annotation: { target: { created: b } } } = a;
      return c.getTime() - b.getTime();
    }).forEach((x) => {
      var y;
      const a = S ? typeof S == "function" ? S(x.annotation, x.state) : S : (y = x.state) != null && y.selected ? X : M, c = p && p.paint(x, g) || a, b = x.rects.map(({ x: m, y: f, width: w, height: T }) => ({
        x: m + C,
        y: f + E,
        width: w,
        height: T
      }));
      if (e.fillStyle = c.fill, e.globalAlpha = c.fillOpacity || 1, b.forEach(
        ({ x: m, y: f, width: w, height: T }) => e.fillRect(m, f, w, T)
      ), c.underlineColor) {
        e.globalAlpha = 1, e.strokeStyle = c.underlineColor, e.lineWidth = c.underlineThickness ?? 1;
        const m = c.underlineOffset ?? 0;
        b.forEach(({ x: f, y: w, width: T, height: R }) => {
          e.beginPath(), e.moveTo(f, w + R + m), e.lineTo(f + T, w + R + m), e.stroke();
        });
      }
    });
  }), s = W(() => {
    Ft(n);
  });
  return window.addEventListener("resize", s), {
    destroy: () => {
      n.remove(), window.removeEventListener("resize", s);
    },
    setVisible: (u) => {
      console.log("setVisible not implemented on Canvas renderer");
    },
    redraw: o
  };
}, Dt = (t, n, e) => G(t, n, e, _t(t)), Vt = (t) => [
  `background-color:${J((t == null ? void 0 : t.fill) || M.fill).alpha((t == null ? void 0 : t.fillOpacity) === void 0 ? M.fillOpacity : t.fillOpacity).toHex()}`,
  t != null && t.underlineThickness ? "text-decoration:underline" : void 0,
  t != null && t.underlineColor ? `text-decoration-color:${t.underlineColor}` : void 0,
  t != null && t.underlineOffset ? `text-underline-offset:${t.underlineOffset}px` : void 0,
  t != null && t.underlineThickness ? `text-decoration-thickness:${t.underlineThickness}px` : void 0
].filter(Boolean).join(";"), Wt = () => {
  const t = document.createElement("style");
  document.getElementsByTagName("head")[0].appendChild(t);
  let n = /* @__PURE__ */ new Set();
  return {
    destroy: () => {
      CSS.highlights.clear(), t.remove();
    },
    setVisible: (r) => {
      console.log("setVisible not implemented on CSS Custom Highlights renderer");
    },
    redraw: (r, l, u, g) => {
      g && g.clear();
      const S = new Set(r.map((d) => d.annotation.id));
      Array.from(n).filter((d) => !S.has(d));
      const p = r.map((d) => {
        var C;
        const i = u ? typeof u == "function" ? u(d.annotation, d.state) : u : (C = d.state) != null && C.selected ? X : M, E = g && g.paint(d, l) || i;
        return `::highlight(_${d.annotation.id}) { ${Vt(E)} }`;
      });
      t.innerHTML = p.join(`
`), CSS.highlights.clear(), r.forEach(({ annotation: d }) => {
        const i = d.target.selector.map((C) => C.range), E = new Highlight(...i);
        CSS.highlights.set(`_${d.id}`, E);
      }), n = S;
    }
  };
}, Yt = (t, n, e) => G(t, n, e, Wt()), Xt = (t, n) => {
  const e = (r, l) => r.x <= l.x + l.width && r.x + r.width >= l.x && r.y <= l.y + l.height && r.y + r.height >= l.y, o = (r) => r.rects.reduce((l, u) => l + u.width, 0), s = n.filter(({ rects: r }) => r.some((l) => e(t, l)));
  return s.sort((r, l) => o(l) - o(r)), s.findIndex((r) => r.rects.includes(t));
}, zt = (t) => {
  t.classList.add("r6o-annotatable");
  const n = document.createElement("div");
  n.className = "r6o-span-highlight-layer", t.insertBefore(n, t.firstChild);
  let e = [];
  return {
    destroy: () => {
      n.remove();
    },
    redraw: (l, u, g, S, p) => {
      const i = !(at(e, l) && p);
      if (!S && !i) return;
      i && (n.innerHTML = ""), [...l].sort((C, v) => {
        const { annotation: { target: { created: x } } } = C, { annotation: { target: { created: a } } } = v;
        return x && a ? x.getTime() - a.getTime() : 0;
      }).forEach((C) => {
        C.rects.map((v) => {
          const x = Xt(v, l), a = Ht(C, u, g, S, x);
          if (i) {
            const c = document.createElement("span");
            c.className = "r6o-annotation", c.dataset.annotation = C.annotation.id, c.style.left = `${v.x}px`, c.style.top = `${v.y}px`, c.style.width = `${v.width}px`, c.style.height = `${v.height}px`, c.style.backgroundColor = J((a == null ? void 0 : a.fill) || M.fill).alpha((a == null ? void 0 : a.fillOpacity) === void 0 ? M.fillOpacity : a.fillOpacity).toHex(), a.underlineStyle && (c.style.borderStyle = a.underlineStyle), a.underlineColor && (c.style.borderColor = a.underlineColor), a.underlineThickness && (c.style.borderBottomWidth = `${a.underlineThickness}px`), a.underlineOffset && (c.style.paddingBottom = `${a.underlineOffset}px`), n.appendChild(c);
          }
        });
      }), e = l;
    },
    setVisible: (l) => {
      l ? n.classList.remove("hidden") : n.classList.add("hidden");
    }
  };
}, qt = (t, n, e) => G(t, n, e, zt(t)), me = (t, n) => ({
  parse: (e) => Qt(e),
  serialize: (e) => Zt(e, t, n)
}), jt = (t) => t.quote !== void 0 && t.start !== void 0 && t.end !== void 0, Gt = (t) => {
  const {
    id: n,
    creator: e,
    created: o,
    modified: s,
    target: r
  } = t, l = Array.isArray(r) ? r : [r];
  if (l.length === 0)
    return { error: Error(`No targets found for annotation: ${t.id}`) };
  const u = {
    creator: dt(e),
    created: o ? new Date(o) : void 0,
    updated: s ? new Date(s) : void 0,
    annotation: n,
    selector: [],
    // @ts-expect-error: `styleClass` is not part of the core `TextAnnotationTarget` type
    styleClass: "styleClass" in l[0] ? l[0].styleClass : void 0
  };
  for (const g of l) {
    const p = (Array.isArray(g.selector) ? g.selector : [g.selector]).reduce((d, i) => {
      switch (i.type) {
        case "TextQuoteSelector":
          d.quote = i.exact;
          break;
        case "TextPositionSelector":
          d.start = i.start, d.end = i.end;
          break;
      }
      return d;
    }, {});
    if (jt(p))
      u.selector.push(
        {
          ...p,
          id: g.id,
          // @ts-expect-error: `scope` is not part of the core `TextSelector` type
          scope: g.scope
        }
      );
    else {
      const d = [
        p.start ? void 0 : "TextPositionSelector",
        p.quote ? void 0 : "TextQuoteSelector"
      ].filter(Boolean);
      return { error: Error(`Missing selector types: ${d.join(" and ")} for annotation: ${t.id}`) };
    }
  }
  return { parsed: u };
}, Qt = (t) => {
  const n = t.id || tt(), {
    creator: e,
    created: o,
    modified: s,
    body: r,
    ...l
  } = t, u = lt(r, n), g = Gt(t);
  return "error" in g ? { error: g.error } : {
    parsed: {
      ...l,
      id: n,
      bodies: u,
      target: g.parsed
    }
  };
}, Zt = (t, n, e) => {
  const { bodies: o, target: s, ...r } = t, {
    selector: l,
    creator: u,
    created: g,
    updated: S,
    ...p
  } = s, d = l.map((i) => {
    const { id: E, quote: C, start: v, end: x, range: a } = i, c = {
      type: "TextQuoteSelector",
      exact: C
    };
    if (e) {
      const { prefix: y, suffix: m } = Rt(a, e);
      c.prefix = y, c.suffix = m;
    }
    const b = {
      type: "TextPositionSelector",
      start: v,
      end: x
    };
    return {
      ...p,
      id: E,
      // @ts-expect-error: `scope` is not part of the core `TextSelector` type
      scope: "scope" in i ? i.scope : void 0,
      source: n,
      selector: [c, b]
    };
  });
  return {
    ...r,
    "@context": "http://www.w3.org/ns/anno.jsonld",
    id: t.id,
    type: "Annotation",
    body: ct(t.bodies),
    creator: u,
    created: g == null ? void 0 : g.toISOString(),
    modified: S == null ? void 0 : S.toISOString(),
    target: d
  };
}, Jt = (t, n) => {
  const e = new St(), o = /* @__PURE__ */ new Map(), s = (a, c) => {
    const b = a.selector.flatMap((m) => {
      const f = P([m]) ? m.range : nt(m, n).range;
      return Array.from(f.getClientRects());
    }), y = Bt(b).map(({ left: m, top: f, right: w, bottom: T }) => new DOMRect(m - c.left, f - c.top, w - m, T - f));
    return y.map((m) => {
      const { x: f, y: w, width: T, height: R } = m;
      return {
        minX: f,
        minY: w,
        maxX: f + T,
        maxY: w + R,
        annotation: {
          id: a.annotation,
          rects: y
        }
      };
    });
  }, r = () => [...o.values()], l = () => {
    e.clear(), o.clear();
  }, u = (a) => {
    const c = s(a, n.getBoundingClientRect());
    c.length !== 0 && (c.forEach((b) => e.insert(b)), o.set(a.annotation, c));
  }, g = (a) => {
    const c = o.get(a.annotation);
    c && (c.forEach((b) => e.remove(b)), o.delete(a.annotation));
  }, S = (a) => {
    g(a), u(a);
  }, p = (a, c = !0) => {
    c && l();
    const b = n.getBoundingClientRect(), y = a.map((f) => ({ target: f, rects: s(f, b) }));
    y.forEach(({ target: f, rects: w }) => {
      w.length > 0 && o.set(f.annotation, w);
    });
    const m = y.flatMap(({ rects: f }) => f);
    e.load(m);
  }, d = (a, c, b = !1) => {
    const y = e.search({
      minX: a,
      minY: c,
      maxX: a,
      maxY: c
    }), m = (f) => f.annotation.rects.reduce((w, T) => w + T.width * T.height, 0);
    return y.length > 0 ? (y.sort((f, w) => m(f) - m(w)), b ? y.map((f) => f.annotation.id) : [y[0].annotation.id]) : [];
  }, i = (a) => {
    const c = E(a);
    if (c.length === 0)
      return;
    let b = c[0].left, y = c[0].top, m = c[0].right, f = c[0].bottom;
    for (let w = 1; w < c.length; w++) {
      const T = c[w];
      b = Math.min(b, T.left), y = Math.min(y, T.top), m = Math.max(m, T.right), f = Math.max(f, T.bottom);
    }
    return new DOMRect(b, y, m - b, f - y);
  }, E = (a) => {
    const c = o.get(a);
    return c ? c[0].annotation.rects : [];
  };
  return {
    all: r,
    clear: l,
    getAt: d,
    getAnnotationBounds: i,
    getAnnotationRects: E,
    getIntersecting: (a, c, b, y) => {
      const m = e.search({ minX: a, minY: c, maxX: b, maxY: y }), f = new Set(m.map((w) => w.annotation.id));
      return Array.from(f).map((w) => ({
        annotation: t.getAnnotation(w),
        rects: E(w)
      })).filter((w) => !!w.annotation);
    },
    insert: u,
    recalculate: () => p(t.all().map((a) => a.target), !0),
    remove: g,
    set: p,
    size: () => e.all().length,
    update: S
  };
}, te = (t, n) => {
  const e = ut(), o = Jt(e, t), s = ft(e, n.userSelectAction, n.adapter), r = gt(e), l = mt(), u = (a, c = K.LOCAL) => {
    const b = q(a, t), y = P(b.target.selector);
    return y && e.addAnnotation(b, c), y;
  }, g = (a, c = !0, b = K.LOCAL) => {
    const y = a.map((f) => q(f, t)), m = y.filter((f) => !P(f.target.selector));
    return e.bulkAddAnnotations(y, c, b), m;
  }, S = (a, c = K.LOCAL) => {
    const b = a.map((m) => q(m, t)), y = b.filter((m) => !P(m.target.selector));
    return b.forEach((m) => {
      e.getAnnotation(m.id) ? e.updateAnnotation(m, c) : e.addAnnotation(m, c);
    }), y;
  }, p = (a, c = K.LOCAL) => {
    const b = Y(a, t);
    e.updateTarget(b, c);
  }, d = (a, c = K.LOCAL) => {
    const b = a.map((y) => Y(y, t));
    e.bulkUpdateTargets(b, c);
  };
  function i(a, c, b, y) {
    const m = b || !!y, f = o.getAt(a, c, m).map((T) => e.getAnnotation(T)), w = y ? f.filter(y) : f;
    if (w.length !== 0)
      return b ? w : w[0];
  }
  const E = (a) => {
    if (o.getAnnotationRects(a).length !== 0)
      return o.getAnnotationBounds(a);
  }, C = (a, c, b, y) => o.getIntersecting(a, c, b, y), v = (a) => o.getAnnotationRects(a), x = () => o.recalculate();
  return e.observe(({ changes: a }) => {
    const c = (a.deleted || []).filter((m) => P(m.target.selector)), b = (a.created || []).filter((m) => P(m.target.selector)), y = (a.updated || []).filter((m) => P(m.newValue.target.selector));
    (c == null ? void 0 : c.length) > 0 && c.forEach((m) => o.remove(m.target)), b.length > 0 && o.set(b.map((m) => m.target), !1), (y == null ? void 0 : y.length) > 0 && y.forEach(({ newValue: m }) => o.update(m.target));
  }), {
    store: {
      ...e,
      addAnnotation: u,
      bulkAddAnnotations: g,
      bulkUpdateTargets: d,
      bulkUpsertAnnotations: S,
      getAnnotationBounds: E,
      getAnnotationRects: v,
      getIntersecting: C,
      getAt: i,
      recalculatePositions: x,
      updateTarget: p
    },
    selection: s,
    hover: r,
    viewport: l
  };
}, ee = () => {
  const t = document.createElement("canvas");
  t.width = 2 * window.innerWidth, t.height = 2 * window.innerHeight, t.className = "r6o-presence-layer";
  const n = t.getContext("2d");
  return n.scale(2, 2), n.translate(0.5, 0.5), t;
}, ne = (t, n = {}) => {
  const e = ee(), o = e.getContext("2d");
  document.body.appendChild(e);
  const s = /* @__PURE__ */ new Map(), r = (p) => Array.from(s.entries()).filter(([d, i]) => i.presenceKey === p.presenceKey).map(([d, i]) => d);
  return t.on("selectionChange", (p, d) => {
    r(p).forEach((E) => s.delete(E)), d && d.forEach((E) => s.set(E, p));
  }), {
    clear: () => {
      const { width: p, height: d } = e;
      o.clearRect(-0.5, -0.5, p + 1, d + 1);
    },
    destroy: () => {
      e.remove();
    },
    paint: (p, d, i) => {
      n.font && (o.font = n.font);
      const E = s.get(p.annotation.id);
      if (E) {
        const { height: C } = p.rects[0], v = p.rects[0].x + d.left, x = p.rects[0].y + d.top;
        o.fillStyle = E.appearance.color, o.fillRect(v - 2, x - 2.5, 2, C + 5);
        const a = o.measureText(E.appearance.label), c = a.width + 6, b = a.actualBoundingBoxAscent + a.actualBoundingBoxDescent + 8, y = a.fontBoundingBoxAscent ? 8 : 6.5;
        return o.fillRect(v - 2, x - 2.5 - b, c, b), o.fillStyle = "#fff", o.fillText(E.appearance.label, v + 1, x - y), {
          fill: E.appearance.color,
          fillOpacity: i ? 0.45 : 0.18
        };
      }
    },
    reset: () => {
      e.width = 2 * window.innerWidth, e.height = 2 * window.innerHeight;
      const p = e.getContext("2d");
      p.scale(2, 2), p.translate(0.5, 0.5);
    }
  };
}, Q = 300, st = ["up", "down", "left", "right"], rt = At ? "⌘+a" : "ctrl+a", oe = [
  ...st.map((t) => `shift+${t}`),
  rt
], se = (t, n, e) => {
  let o;
  const { annotatingEnabled: s, offsetReferenceSelector: r, selectionMode: l } = e, u = (h) => o = h;
  let g;
  const S = (h) => g = h, { store: p, selection: d } = n;
  let i, E, C;
  const v = (h) => {
    E !== !1 && (i = F(t, h.target) ? void 0 : {
      annotation: tt(),
      selector: [],
      creator: o,
      created: /* @__PURE__ */ new Date()
    });
  }, x = W((h) => {
    const A = document.getSelection();
    if (!(A != null && A.anchorNode))
      return;
    if (F(t, A.anchorNode)) {
      i = void 0;
      return;
    }
    const L = h.timeStamp - ((C == null ? void 0 : C.timeStamp) || h.timeStamp);
    if ((C == null ? void 0 : C.type) === "pointerdown" && (L < 1e3 && !i || A.isCollapsed && L < Q) && v(C || h), !i) return;
    if (A.isCollapsed) {
      p.getAnnotation(i.annotation) && (d.clear(), p.deleteAnnotation(i.annotation));
      return;
    }
    const N = Array.from(Array(A.rangeCount).keys()).map((k) => A.getRangeAt(k)).map((k) => Pt(k, t));
    if (N.every((k) => kt(k))) return;
    const I = N.flatMap((k) => Tt(t, k.cloneRange()));
    (I.length !== i.selector.length || I.some((k, O) => {
      var $;
      return k.toString() !== (($ = i.selector[O]) == null ? void 0 : $.quote);
    })) && (i = {
      ...i,
      selector: I.map((k) => Nt(k, t, r)),
      updated: /* @__PURE__ */ new Date()
    }, p.getAnnotation(i.annotation) ? p.updateTarget(i, K.LOCAL) : d.clear());
  }), a = (h) => {
    F(t, h.target) || (C = z(h), E = C.button === 0);
  }, c = async (h) => {
    if (!E) return;
    if (F(t, h.target)) {
      e.dismissOnClickOutside && d.clear();
      return;
    }
    const A = () => {
      const { x: B, y: N } = t.getBoundingClientRect(), I = h.target instanceof Node && t.contains(h.target) && p.getAt(h.clientX - B, h.clientY - N, l === "all", g);
      if (I) {
        const { selected: _ } = d, k = new Set(_.map((H) => H.id)), O = Array.isArray(I) ? I.map((H) => H.id) : [I.id];
        (k.size !== O.length || !O.every((H) => k.has(H))) && d.userSelect(O, h);
      } else
        d.clear();
    };
    if (h.timeStamp - C.timeStamp < Q) {
      await b();
      const B = document.getSelection();
      if (B != null && B.isCollapsed) {
        i = void 0, A();
        return;
      }
    }
    i && i.selector.length > 0 && (T(), d.userSelect(i.annotation, z(h)));
  }, b = async () => {
    const h = document.getSelection();
    let A = !1, L = h == null ? void 0 : h.isCollapsed;
    const B = () => L || A, N = 1;
    return setTimeout(() => A = !0, 50), Ct(() => L = h == null ? void 0 : h.isCollapsed, N, B);
  }, y = (h) => {
    const A = document.getSelection();
    A != null && A.isCollapsed || ((!i || i.selector.length === 0) && x(h), i && (T(), d.userSelect(i.annotation, z(h))));
  }, m = (h) => {
    h.key === "Shift" && i && (document.getSelection().isCollapsed || (T(), d.userSelect(i.annotation, V(h))));
  }, f = (h) => {
    const A = () => setTimeout(() => {
      (i == null ? void 0 : i.selector.length) > 0 && (d.clear(), p.addAnnotation({
        id: i.annotation,
        bodies: [],
        target: i
      }), d.userSelect(i.annotation, V(h))), document.removeEventListener("selectionchange", A);
    }, 100);
    document.addEventListener("selectionchange", A), v(h);
  };
  D(oe.join(","), { element: t, keydown: !0, keyup: !1 }, (h) => {
    h.repeat || (C = V(h));
  }), D(rt, { keydown: !0, keyup: !1 }, (h) => {
    C = V(h), f(h);
  });
  const w = (h) => {
    h.repeat || h.target !== t && h.target !== document.body || (i = void 0, d.clear());
  };
  D(st.join(","), { keydown: !0, keyup: !1 }, w);
  const T = () => {
    const h = p.getAnnotation(i.annotation);
    if (!h) {
      p.addAnnotation({
        id: i.annotation,
        bodies: [],
        target: i
      });
      return;
    }
    const { target: { updated: A } } = h, { updated: L } = i;
    (!A || !L || A < L) && p.updateTarget(i);
  };
  return t.addEventListener("pointerdown", a), document.addEventListener("pointerup", c), document.addEventListener("contextmenu", y), s && (t.addEventListener("keyup", m), t.addEventListener("selectstart", v), document.addEventListener("selectionchange", x)), {
    destroy: () => {
      t.removeEventListener("pointerdown", a), document.removeEventListener("pointerup", c), document.removeEventListener("contextmenu", y), t.removeEventListener("keyup", m), t.removeEventListener("selectstart", v), document.removeEventListener("selectionchange", x), D.unbind();
    },
    setFilter: S,
    setUser: u
  };
}, re = (t, n) => ({
  ...t,
  annotatingEnabled: t.annotatingEnabled ?? n.annotatingEnabled,
  user: t.user || n.user
}), Z = "SPANS", pe = (t, n = {}) => {
  Et(t), xt(t);
  const e = re(n, {
    annotatingEnabled: !0,
    user: pt()
  }), o = te(t, e), { selection: s, viewport: r } = o, l = o.store, u = ht(l), g = yt(o, u, e.adapter);
  let S = e.user;
  const p = e.renderer === "CSS_HIGHLIGHTS" ? CSS.highlights ? "CSS_HIGHLIGHTS" : Z : e.renderer || Z, d = p === "SPANS" ? qt(t, o, r) : p === "CSS_HIGHLIGHTS" ? Yt(t, o, r) : p === "CANVAS" ? Dt(t, o, r) : void 0;
  if (!d)
    throw `Unknown renderer implementation: ${p}`;
  console.debug(`Using ${p} renderer`), e.style && d.setStyle(e.style);
  const i = se(t, o, e);
  return i.setUser(S), {
    ...bt(o, u, e.adapter),
    destroy: () => {
      d.destroy(), i.destroy(), u.destroy();
    },
    element: t,
    getUser: () => S,
    setFilter: (f) => {
      d.setFilter(f), i.setFilter(f);
    },
    setStyle: (f) => d.setStyle(f),
    setUser: (f) => {
      S = f, i.setUser(f);
    },
    setSelected: (f) => {
      f ? s.setSelected(f) : s.clear();
    },
    setPresenceProvider: (f) => {
      f && (d.setPainter(ne(f, e.presence)), f.on("selectionChange", () => d.redraw()));
    },
    setVisible: (f) => d.setVisible(f),
    on: g.on,
    off: g.off,
    scrollIntoView: Mt(t, l),
    state: o
  };
};
export {
  X as DEFAULT_SELECTED_STYLE,
  M as DEFAULT_STYLE,
  et as NOT_ANNOTATABLE_CLASS,
  U as NOT_ANNOTATABLE_SELECTOR,
  be as Origin,
  se as SelectionHandler,
  Se as UserSelectAction,
  me as W3CTextFormat,
  Et as cancelSingleClickEvents,
  V as cloneKeyboardEvent,
  z as clonePointerEvent,
  Ce as createBody,
  Dt as createCanvasRenderer,
  Yt as createHighlightsRenderer,
  ne as createPresencePainter,
  Wt as createRenderer,
  qt as createSpansRenderer,
  pe as createTextAnnotator,
  te as createTextAnnotatorState,
  W as debounce,
  re as fillDefaults,
  Rt as getQuoteContext,
  j as getRangeAnnotatableContents,
  At as isMac,
  F as isNotAnnotatable,
  wt as isRangeAnnotatable,
  P as isRevived,
  kt as isWhitespaceOrEmpty,
  Bt as mergeClientRects,
  Ht as paint,
  Qt as parseW3CTextAnnotation,
  xt as programmaticallyFocusable,
  Nt as rangeToSelector,
  q as reviveAnnotation,
  nt as reviveSelector,
  Y as reviveTarget,
  Mt as scrollIntoView,
  Zt as serializeW3CTextAnnotation,
  Tt as splitAnnotatableRanges,
  ge as toDomRectList,
  Pt as trimRangeToContainer,
  Lt as whitespaceOrEmptyRegex
};
