// @annotorious/react/dist/annotorious-react.es.js
import { Annotorious as r, AnnotoriousContext as a, useAnnotation as p, useAnnotationSelectAction as i, useAnnotationStore as u, useAnnotations as s, useAnnotator as m, useAnnotatorUser as A, useHover as S, useSelection as x, useViewportState as c } from "./annotorious-react.es2.js";
import { AnnotoriousPlugin as g } from "./annotorious-react.es3.js";
import { ImageAnnotationPopup as d } from "./annotorious-react.es4.js";
import { ImageAnnotator as O } from "./annotorious-react.es5.js";
/* empty css                      */
/* empty css                      */
import { Origin as I, UserSelectAction as P, createAnonymousGuest as w, createBody as V, defaultColorProvider as h } from "@annotorious/core";
import { ShapeType as U, W3CImageFormat as B, chainStyles as F, computeStyle as G, createImageAnnotator as H } from "@annotorious/annotorious";
import { OpenSeadragonAnnotator as W, OpenSeadragonAnnotatorContext as b, useViewer as j } from "./annotorious-react.es8.js";
import { OpenSeadragonAnnotationPopup as q } from "./annotorious-react.es9.js";
import { OpenSeadragonViewer as D } from "./annotorious-react.es10.js";
export {
  r as Annotorious,
  a as AnnotoriousContext,
  g as AnnotoriousPlugin,
  d as ImageAnnotationPopup,
  O as ImageAnnotator,
  q as OpenSeadragonAnnotationPopup,
  W as OpenSeadragonAnnotator,
  b as OpenSeadragonAnnotatorContext,
  D as OpenSeadragonViewer,
  I as Origin,
  U as ShapeType,
  P as UserSelectAction,
  B as W3CImageFormat,
  F as chainStyles,
  G as computeStyle,
  w as createAnonymousGuest,
  V as createBody,
  H as createImageAnnotator,
  h as defaultColorProvider,
  p as useAnnotation,
  i as useAnnotationSelectAction,
  u as useAnnotationStore,
  s as useAnnotations,
  m as useAnnotator,
  A as useAnnotatorUser,
  S as useHover,
  x as useSelection,
  j as useViewer,
  c as useViewportState
};
