// @recogito/text-annotator-tei/dist/text-annotator-tei.es.js
const Q = [];
for (let e = 0; e < 256; ++e)
  Q.push((e + 256).toString(16).slice(1));
typeof crypto < "u" && crypto.randomUUID && crypto.randomUUID.bind(crypto);
const J = "useandom-26T198340PX75pxJACKVERYMINDBUSHWOLF_GQZbfghjklqvwyzrict";
let Z = (e = 21) => {
  let t = "", n = crypto.getRandomValues(new Uint8Array(e |= 0));
  for (; e--; )
    t += J[n[e] & 63];
  return t;
};
const ee = (e) => e.map((t) => {
  var n;
  const { annotation: r, created: o, updated: i, ...c } = t, s = {
    ...c,
    created: o == null ? void 0 : o.toISOString(),
    modified: i == null ? void 0 : i.toISOString()
  };
  return (n = s.id) != null && n.startsWith("temp-") && delete s.id, s;
});
Z();
const Ce = (e) => ({
  parse: (t) => te(),
  serialize: (t) => ne(t, e)
}), te = (e) => {
}, ne = (e, t) => {
  const { bodies: n, target: r, ...o } = e, {
    selector: i,
    creator: c,
    created: s,
    updated: f,
    ...a
  } = r, l = i.map((d) => {
    const { id: u, quote: p, startSelector: w, endSelector: h } = d;
    return {
      ...a,
      id: u,
      source: t,
      selector: [{
        type: "TextQuoteSelector",
        exact: p
      }, {
        type: "RangeSelector",
        startSelector: w,
        endSelector: h
      }]
    };
  });
  return {
    ...o,
    "@context": "http://www.w3.org/ns/anno.jsonld",
    id: e.id,
    type: "Annotation",
    body: ee(e.bodies),
    creator: c,
    created: s == null ? void 0 : s.toISOString(),
    modified: f == null ? void 0 : f.toISOString(),
    target: l
  };
}, I = typeof navigator < "u" ? navigator.userAgent.toLowerCase().indexOf("firefox") > 0 : !1;
function P(e, t, n, r) {
  e.addEventListener ? e.addEventListener(t, n, r) : e.attachEvent && e.attachEvent("on".concat(t), n);
}
function C(e, t, n, r) {
  e.removeEventListener ? e.removeEventListener(t, n, r) : e.detachEvent && e.detachEvent("on".concat(t), n);
}
function U(e, t) {
  const n = t.slice(0, t.length - 1);
  for (let r = 0; r < n.length; r++) n[r] = e[n[r].toLowerCase()];
  return n;
}
function X(e) {
  typeof e != "string" && (e = ""), e = e.replace(/\s/g, "");
  const t = e.split(",");
  let n = t.lastIndexOf("");
  for (; n >= 0; )
    t[n - 1] += ",", t.splice(n, 1), n = t.lastIndexOf("");
  return t;
}
function oe(e, t) {
  const n = e.length >= t.length ? e : t, r = e.length >= t.length ? t : e;
  let o = !0;
  for (let i = 0; i < n.length; i++)
    r.indexOf(n[i]) === -1 && (o = !1);
  return o;
}
const K = {
  backspace: 8,
  "⌫": 8,
  tab: 9,
  clear: 12,
  enter: 13,
  "↩": 13,
  return: 13,
  esc: 27,
  escape: 27,
  space: 32,
  left: 37,
  up: 38,
  right: 39,
  down: 40,
  /// https://w3c.github.io/uievents/#events-keyboard-key-location
  arrowup: 38,
  arrowdown: 40,
  arrowleft: 37,
  arrowright: 39,
  del: 46,
  delete: 46,
  ins: 45,
  insert: 45,
  home: 36,
  end: 35,
  pageup: 33,
  pagedown: 34,
  capslock: 20,
  num_0: 96,
  num_1: 97,
  num_2: 98,
  num_3: 99,
  num_4: 100,
  num_5: 101,
  num_6: 102,
  num_7: 103,
  num_8: 104,
  num_9: 105,
  num_multiply: 106,
  num_add: 107,
  num_enter: 108,
  num_subtract: 109,
  num_decimal: 110,
  num_divide: 111,
  "⇪": 20,
  ",": 188,
  ".": 190,
  "/": 191,
  "`": 192,
  "-": I ? 173 : 189,
  "=": I ? 61 : 187,
  ";": I ? 59 : 186,
  "'": 222,
  "{": 219,
  "}": 221,
  "[": 219,
  "]": 221,
  "\\": 220
}, E = {
  // shiftKey
  "⇧": 16,
  shift: 16,
  // altKey
  "⌥": 18,
  alt: 18,
  option: 18,
  // ctrlKey
  "⌃": 17,
  ctrl: 17,
  control: 17,
  // metaKey
  "⌘": 91,
  cmd: 91,
  meta: 91,
  command: 91
}, T = {
  16: "shiftKey",
  18: "altKey",
  17: "ctrlKey",
  91: "metaKey",
  shiftKey: 16,
  ctrlKey: 17,
  altKey: 18,
  metaKey: 91
}, m = {
  16: !1,
  18: !1,
  17: !1,
  91: !1
}, y = {};
for (let e = 1; e < 20; e++)
  K["f".concat(e)] = 111 + e;
let g = [], A = null, W = "all";
const S = /* @__PURE__ */ new Map(), b = (e) => K[e.toLowerCase()] || E[e.toLowerCase()] || e.toUpperCase().charCodeAt(0), re = (e) => Object.keys(K).find((t) => K[t] === e), se = (e) => Object.keys(E).find((t) => E[t] === e);
function q(e) {
  W = e || "all";
}
function _() {
  return W || "all";
}
function ie() {
  return g.slice(0);
}
function ce() {
  return g.map((e) => re(e) || se(e) || String.fromCharCode(e));
}
function ae() {
  const e = [];
  return Object.keys(y).forEach((t) => {
    y[t].forEach((n) => {
      let {
        key: r,
        scope: o,
        mods: i,
        shortcut: c
      } = n;
      e.push({
        scope: o,
        shortcut: c,
        mods: i,
        keys: r.split("+").map((s) => b(s))
      });
    });
  }), e;
}
function le(e) {
  const t = e.target || e.srcElement, {
    tagName: n
  } = t;
  let r = !0;
  const o = n === "INPUT" && !["checkbox", "radio", "range", "button", "file", "reset", "submit", "color"].includes(t.type);
  return (t.isContentEditable || (o || n === "TEXTAREA" || n === "SELECT") && !t.readOnly) && (r = !1), r;
}
function fe(e) {
  return typeof e == "string" && (e = b(e)), g.indexOf(e) !== -1;
}
function ue(e, t) {
  let n, r;
  e || (e = _());
  for (const o in y)
    if (Object.prototype.hasOwnProperty.call(y, o))
      for (n = y[o], r = 0; r < n.length; )
        n[r].scope === e ? n.splice(r, 1).forEach((c) => {
          let {
            element: s
          } = c;
          return j(s);
        }) : r++;
  _() === e && q(t || "all");
}
function de(e) {
  let t = e.keyCode || e.which || e.charCode;
  e.key && e.key.toLowerCase() === "capslock" && (t = b(e.key));
  const n = g.indexOf(t);
  if (n >= 0 && g.splice(n, 1), e.key && e.key.toLowerCase() === "meta" && g.splice(0, g.length), (t === 93 || t === 224) && (t = 91), t in m) {
    m[t] = !1;
    for (const r in E) E[r] === t && (x[r] = !1);
  }
}
function z(e) {
  if (typeof e > "u")
    Object.keys(y).forEach((o) => {
      Array.isArray(y[o]) && y[o].forEach((i) => L(i)), delete y[o];
    }), j(null);
  else if (Array.isArray(e))
    e.forEach((o) => {
      o.key && L(o);
    });
  else if (typeof e == "object")
    e.key && L(e);
  else if (typeof e == "string") {
    for (var t = arguments.length, n = new Array(t > 1 ? t - 1 : 0), r = 1; r < t; r++)
      n[r - 1] = arguments[r];
    let [o, i] = n;
    typeof o == "function" && (i = o, o = ""), L({
      key: e,
      scope: o,
      method: i,
      splitKey: "+"
    });
  }
}
const L = (e) => {
  let {
    key: t,
    scope: n,
    method: r,
    splitKey: o = "+"
  } = e;
  X(t).forEach((c) => {
    const s = c.split(o), f = s.length, a = s[f - 1], l = a === "*" ? "*" : b(a);
    if (!y[l]) return;
    n || (n = _());
    const d = f > 1 ? U(E, s) : [], u = [];
    y[l] = y[l].filter((p) => {
      const h = (r ? p.method === r : !0) && p.scope === n && oe(p.mods, d);
      return h && u.push(p.element), !h;
    }), u.forEach((p) => j(p));
  });
};
function R(e, t, n, r) {
  if (t.element !== r)
    return;
  let o;
  if (t.scope === n || t.scope === "all") {
    o = t.mods.length > 0;
    for (const i in m)
      Object.prototype.hasOwnProperty.call(m, i) && (!m[i] && t.mods.indexOf(+i) > -1 || m[i] && t.mods.indexOf(+i) === -1) && (o = !1);
    (t.mods.length === 0 && !m[16] && !m[18] && !m[17] && !m[91] || o || t.shortcut === "*") && (t.keys = [], t.keys = t.keys.concat(g), t.method(e, t) === !1 && (e.preventDefault ? e.preventDefault() : e.returnValue = !1, e.stopPropagation && e.stopPropagation(), e.cancelBubble && (e.cancelBubble = !0)));
  }
}
function F(e, t) {
  const n = y["*"];
  let r = e.keyCode || e.which || e.charCode;
  if (e.key && e.key.toLowerCase() === "capslock" || !x.filter.call(this, e)) return;
  if ((r === 93 || r === 224) && (r = 91), g.indexOf(r) === -1 && r !== 229 && g.push(r), ["metaKey", "ctrlKey", "altKey", "shiftKey"].forEach((s) => {
    const f = T[s];
    e[s] && g.indexOf(f) === -1 ? g.push(f) : !e[s] && g.indexOf(f) > -1 ? g.splice(g.indexOf(f), 1) : s === "metaKey" && e[s] && (g = g.filter((a) => a in T || a === r));
  }), r in m) {
    m[r] = !0;
    for (const s in E)
      if (Object.prototype.hasOwnProperty.call(E, s)) {
        const f = T[E[s]];
        x[s] = e[f];
      }
    if (!n) return;
  }
  for (const s in m)
    Object.prototype.hasOwnProperty.call(m, s) && (m[s] = e[T[s]]);
  e.getModifierState && !(e.altKey && !e.ctrlKey) && e.getModifierState("AltGraph") && (g.indexOf(17) === -1 && g.push(17), g.indexOf(18) === -1 && g.push(18), m[17] = !0, m[18] = !0);
  const o = _();
  if (n)
    for (let s = 0; s < n.length; s++)
      n[s].scope === o && (e.type === "keydown" && n[s].keydown || e.type === "keyup" && n[s].keyup) && R(e, n[s], o, t);
  if (!(r in y)) return;
  const i = y[r], c = i.length;
  for (let s = 0; s < c; s++)
    if ((e.type === "keydown" && i[s].keydown || e.type === "keyup" && i[s].keyup) && i[s].key) {
      const f = i[s], {
        splitKey: a
      } = f, l = f.key.split(a), d = [];
      for (let u = 0; u < l.length; u++)
        d.push(b(l[u]));
      d.sort().join("") === g.sort().join("") && R(e, f, o, t);
    }
}
function x(e, t, n) {
  g = [];
  const r = X(e);
  let o = [], i = "all", c = document, s = 0, f = !1, a = !0, l = "+", d = !1, u = !1;
  for (n === void 0 && typeof t == "function" && (n = t), Object.prototype.toString.call(t) === "[object Object]" && (t.scope && (i = t.scope), t.element && (c = t.element), t.keyup && (f = t.keyup), t.keydown !== void 0 && (a = t.keydown), t.capture !== void 0 && (d = t.capture), typeof t.splitKey == "string" && (l = t.splitKey), t.single === !0 && (u = !0)), typeof t == "string" && (i = t), u && z(e, i); s < r.length; s++)
    e = r[s].split(l), o = [], e.length > 1 && (o = U(E, e)), e = e[e.length - 1], e = e === "*" ? "*" : b(e), e in y || (y[e] = []), y[e].push({
      keyup: f,
      keydown: a,
      scope: i,
      mods: o,
      shortcut: r[s],
      method: n,
      key: r[s],
      splitKey: l,
      element: c
    });
  if (typeof c < "u" && window) {
    if (!S.has(c)) {
      const p = function() {
        let h = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : window.event;
        return F(h, c);
      }, w = function() {
        let h = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : window.event;
        F(h, c), de(h);
      };
      S.set(c, {
        keydownListener: p,
        keyupListenr: w,
        capture: d
      }), P(c, "keydown", p, d), P(c, "keyup", w, d);
    }
    if (!A) {
      const p = () => {
        g = [];
      };
      A = {
        listener: p,
        capture: d
      }, P(window, "focus", p, d);
    }
  }
}
function pe(e) {
  let t = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : "all";
  Object.keys(y).forEach((n) => {
    y[n].filter((o) => o.scope === t && o.shortcut === e).forEach((o) => {
      o && o.method && o.method();
    });
  });
}
function j(e) {
  const t = Object.values(y).flat();
  if (t.findIndex((r) => {
    let {
      element: o
    } = r;
    return o === e;
  }) < 0) {
    const {
      keydownListener: r,
      keyupListenr: o,
      capture: i
    } = S.get(e) || {};
    r && o && (C(e, "keyup", o, i), C(e, "keydown", r, i), S.delete(e));
  }
  if ((t.length <= 0 || S.size <= 0) && (Object.keys(S).forEach((o) => {
    const {
      keydownListener: i,
      keyupListenr: c,
      capture: s
    } = S.get(o) || {};
    i && c && (C(o, "keyup", c, s), C(o, "keydown", i, s), S.delete(o));
  }), S.clear(), Object.keys(y).forEach((o) => delete y[o]), A)) {
    const {
      listener: o,
      capture: i
    } = A;
    C(window, "focus", o, i), A = null;
  }
}
const v = {
  getPressedKeyString: ce,
  setScope: q,
  getScope: _,
  deleteScope: ue,
  getPressedKeyCodes: ie,
  getAllKeyCodes: ae,
  isPressed: fe,
  filter: le,
  trigger: pe,
  unbind: z,
  keyMap: K,
  modifier: E,
  modifierMap: T
};
for (const e in v)
  Object.prototype.hasOwnProperty.call(v, e) && (x[e] = v[e]);
if (typeof window < "u") {
  const e = window.hotkeys;
  x.noConflict = (t) => (t && window.hotkeys === x && (window.hotkeys = e), x), window.hotkeys = x;
}
const ge = "not-annotatable", B = `.${ge}`;
typeof navigator < "u" && // @ts-ignore
/mac/i.test(navigator.userAgentData ? navigator.userAgentData.platform : navigator.platform);
const ye = (e) => {
  const t = e.cloneContents();
  return t.querySelectorAll(B).forEach((n) => n.remove()), t;
}, he = (e) => e.every((t) => t.range instanceof Range && !t.range.collapsed), me = (e, t, n) => {
  const r = document.createRange(), o = t;
  r.setStart(o, 0), r.setEnd(e.startContainer, e.startOffset);
  const i = ye(r).textContent, c = e.toString(), s = i.length || 0, f = s + c.length;
  return { quote: c, start: s, end: f, range: e };
}, we = (e, t) => {
  var n, r;
  const { start: o, end: i } = e, c = e.offsetReference || t, s = document.createNodeIterator(
    t,
    NodeFilter.SHOW_TEXT,
    (u) => {
      var p;
      return (p = u.parentElement) != null && p.closest(B) ? NodeFilter.FILTER_SKIP : NodeFilter.FILTER_ACCEPT;
    }
  );
  let f = 0;
  const a = document.createRange();
  let l = s.nextNode();
  l === null && console.error("Could not revive annotation target. Content missing.");
  let d = !c;
  for (; l !== null; ) {
    if (d || (d = typeof (c == null ? void 0 : c.contains) == "function" ? c.contains(l) : !1), d) {
      const u = ((n = l.textContent) == null ? void 0 : n.length) || 0;
      if (f + u > o) {
        a.setStart(l, o - f);
        break;
      }
      f += u;
    }
    l = s.nextNode();
  }
  for (; l !== null; ) {
    const u = ((r = l.textContent) == null ? void 0 : r.length) || 0;
    if (f + u >= i) {
      a.setEnd(l, i - f);
      break;
    }
    f += u, l = s.nextNode();
  }
  return {
    ...e,
    range: a
  };
}, D = (e, t) => he(e.selector) ? e : {
  ...e,
  selector: e.selector.map((n) => n.range instanceof Range && !n.range.collapsed ? n : we(n, t))
}, Ee = (e, t, n) => {
  let r = e, o = n;
  const i = document.createNodeIterator(t, NodeFilter.SHOW_TEXT);
  let c = i.nextNode(), s = !0;
  do
    c instanceof Text && (c.length < o ? o -= c.length : (r = c, s = !1)), c = i.nextNode();
  while (c && s);
  return { node: r, offset: o };
}, M = (e, t = []) => {
  let n, r, o;
  if (e.nodeType === Node.ELEMENT_NODE && e.hasAttribute("xml:id") ? t.push("/") : e.parentNode && (t = M(e.parentNode, t)), e.nodeType === Node.ELEMENT_NODE && e.nodeName.toLowerCase().startsWith("tei-")) {
    const i = e;
    i.hasAttribute("xml:id") ? o = `[@xml:id='${i.getAttribute("xml:id")}']` : (n = `count(preceding-sibling::${i.localName})`, r = document.evaluate(n, e, null, XPathResult.NUMBER_TYPE, null).numberValue + 1, o = `[${r}]`), t.push("/"), t.push(i.getAttribute("data-origname") + o);
  }
  return t;
}, Oe = (e, t, n) => {
  const r = (a) => a ? a.nodeName.toLowerCase().indexOf("tei-") === 0 ? a : r(a.parentNode) : null, o = (a, l, d) => {
    const u = document.createRange();
    return u.setStart(a, 0), u.setEnd(l, d), u.toString().length;
  }, i = o(
    r(n.startContainer),
    n.startContainer,
    n.startOffset
  ), c = o(
    r(n.endContainer),
    n.endContainer,
    n.endOffset
  ), s = e.join("") + "::" + i, f = t.join("") + "::" + c;
  return { start: s, end: f };
}, Se = (e) => {
  var c;
  const { range: t } = e, n = M(t.startContainer), r = M(t.endContainer), { start: o, end: i } = Oe(n, r, t);
  return {
    start: e.start,
    startSelector: {
      type: "XPathSelector",
      value: o
    },
    end: e.end,
    endSelector: {
      type: "XPathSelector",
      value: i
    },
    quote: (c = e.quote) == null ? void 0 : c.replace(/\s+/g, " "),
    range: t
  };
}, xe = (e, t) => {
  var r, o;
  const n = Array.isArray(e.selector) ? e.selector[0] : e.selector;
  if ("start" in n && "end" in n)
    return D(e, t);
  {
    const i = (r = n.startSelector) == null ? void 0 : r.value, c = (o = n.endSelector) == null ? void 0 : o.value;
    if (!i || !c)
      throw console.error(e), "Could not revive TEI target.";
    const s = (O) => {
      const k = O.indexOf("::");
      if (k < 0) return;
      const V = O.substring(0, k).replace(/\/([^[/]+)/g, (be, G) => "/tei-" + G.toLowerCase()).replace(/xml:/g, ""), $ = document.evaluate(
        "." + V,
        t,
        null,
        XPathResult.FIRST_ORDERED_NODE_TYPE,
        null
      ).singleNodeValue, Y = parseInt(O.substring(k + 2));
      return [$, Y];
    }, [f, a] = s(i), [l, d] = s(c), u = document.createRange(), p = (O, k) => O.firstChild instanceof Text && O.firstChild.length >= k ? { node: O.firstChild, offset: k } : Ee(O.firstChild, O, k), w = p(f, a);
    u.setStart(w.node, w.offset);
    const h = p(l, d);
    u.setEnd(h.node, h.offset);
    const N = me(u, t);
    return D({
      ...e,
      selector: [{
        ...N,
        ...n,
        range: u
      }]
    }, t);
  }
}, H = (e) => (t) => {
  const n = xe(t, e);
  return {
    ...t,
    selector: n.selector.map(Se)
  };
}, ke = (e) => (t) => ({
  ...t,
  target: H(e)(t.target)
}), Te = (e) => {
  const t = e.element, n = ke(t), r = H(t), o = e.state.store, i = o.addAnnotation;
  o.addAnnotation = (a, l) => {
    const { selector: d } = a.target;
    try {
      return "startSelector" in d && "start" in d ? i(a, l) : i(n(a), l);
    } catch (u) {
      console.warn(u), console.warn("Failed to render annotation", a);
    }
  };
  const c = o.bulkAddAnnotations;
  o.bulkAddAnnotations = (a, l = !0, d) => {
    const u = a.map((w) => {
      const { selector: h } = w.target;
      try {
        return "startSelector" in h && "start" in h ? w : n(w);
      } catch (N) {
        console.warn(N);
      }
    }), p = u.filter(Boolean);
    return u.length !== p.length && (console.warn(`Failed to render ${u.length - p.length} annotations.`), u.forEach((w, h) => {
      w || console.warn(a[h]);
    })), c(p, l, d);
  };
  const s = o.updateAnnotation;
  o.updateAnnotation = (a, l) => s(n(a), l);
  const f = o.updateTarget;
  return o.updateTarget = (a, l) => f(r(a), l), {
    ...e,
    state: {
      ...e.state,
      // @ts-ignore
      store: o
    }
  };
};
export {
  Te as TEIPlugin,
  Ce as W3CTextFormat,
  te as parseW3CTEIAnnotation,
  ne as serializeW3CTEIAnnotation
};
