// @annotorious/react/dist/annotorious-react.es10.js
import { jsx as b } from "react/jsx-runtime";
import { useState as c, useEffect as E } from "react";
import { createPortal as h } from "react-dom";
import { useAnnotator as y } from "./annotorious-react.es2.js";
import { useViewer as A } from "./annotorious-react.es8.js";
const m = 10, T = (n) => {
  const a = n.offsetX || m, v = n.offsetY || m, o = y(), t = A(), [s, p] = c(), [r, i] = c();
  return E(() => {
    if (!o || !t) return;
    const l = (e) => {
      i(e ? o.state.store.getAnnotation(e) : void 0);
    }, f = (e) => {
      const { offsetX: d, offsetY: x } = e;
      p({ x: d, y: x });
    };
    t.element.addEventListener("pointermove", f);
    const u = o.state.hover.subscribe(l);
    return () => {
      var e;
      (e = t.element) == null || e.removeEventListener("pointermove", f), u();
    };
  }, [o, t]), r && s && h(
    /* @__PURE__ */ b(
      "div",
      {
        style: {
          position: "absolute",
          top: `${s.y + v}px`,
          left: `${s.x + a}px`
        },
        children: n.tooltip({ annotation: r })
      }
    ),
    t.element
  );
};
export {
  T as OpenSeadragonHoverTooltip
};
