// @annotorious/react/dist/annotorious-react.es.js
import { Annotorious as r, AnnotoriousContext as a, useAnnotation as i, useAnnotationSelectAction as p, useAnnotationStore as s, useAnnotations as m, useAnnotator as u, useAnnotatorUser as A, useHover as S, useSelection as l, useViewportState as g } from "./annotorious-react.es2.js";
import { AnnotoriousPlugin as x } from "./annotorious-react.es3.js";
import { ImageAnnotationPopup as d } from "./annotorious-react.es4.js";
import { ImageAnnotator as I } from "./annotorious-react.es5.js";
/* empty css                      */
/* empty css                      */
import { Origin as y, UserSelectAction as P, createAnonymousGuest as V, createBody as F, defaultColorProvider as v } from "@annotorious/core";
import { ShapeType as z, W3CImageFormat as G, boundsFromPoints as W, chainStyles as h, computeStyle as H, createImageAnnotator as T, parseFragmentSelector as U, parseSVGSelector as b, parseW3CImageAnnotation as B, serializeFragmentSelector as j, serializeSVGSelector as k, serializeW3CImageAnnotation as q } from "@annotorious/annotorious";
import { OpenSeadragonAnnotator as E, OpenSeadragonAnnotatorContext as J, useViewer as K } from "./annotorious-react.es8.js";
import { OpenSeadragonAnnotationPopup as M } from "./annotorious-react.es9.js";
import { OpenSeadragonHoverTooltip as Q } from "./annotorious-react.es10.js";
import { OpenSeadragonViewer as X } from "./annotorious-react.es11.js";
export {
  r as Annotorious,
  a as AnnotoriousContext,
  x as AnnotoriousPlugin,
  d as ImageAnnotationPopup,
  I as ImageAnnotator,
  M as OpenSeadragonAnnotationPopup,
  E as OpenSeadragonAnnotator,
  J as OpenSeadragonAnnotatorContext,
  Q as OpenSeadragonHoverTooltip,
  X as OpenSeadragonViewer,
  y as Origin,
  z as ShapeType,
  P as UserSelectAction,
  G as W3CImageFormat,
  W as boundsFromPoints,
  h as chainStyles,
  H as computeStyle,
  V as createAnonymousGuest,
  F as createBody,
  T as createImageAnnotator,
  v as defaultColorProvider,
  U as parseFragmentSelector,
  b as parseSVGSelector,
  B as parseW3CImageAnnotation,
  j as serializeFragmentSelector,
  k as serializeSVGSelector,
  q as serializeW3CImageAnnotation,
  i as useAnnotation,
  p as useAnnotationSelectAction,
  s as useAnnotationStore,
  m as useAnnotations,
  u as useAnnotator,
  A as useAnnotatorUser,
  S as useHover,
  l as useSelection,
  K as useViewer,
  g as useViewportState
};
