// @annotorious/react/dist/annotorious-react.es9.js
import { jsxs as h, jsx as O } from "react/jsx-runtime";
import { useState as P, useRef as A, useEffect as D } from "react";
import { createPortal as L } from "react-dom";
import g from "openseadragon";
import { useAnnotator as S, useSelection as b } from "./annotorious-react.es2.js";
/* empty css                      */
/* empty css                      */
import "@annotorious/core";
import "@annotorious/annotorious";
import { toClientRects as U } from "./annotorious-react.es14.js";
import { useFloating as F, FloatingArrow as M } from "./annotorious-react.es13.js";
import { useViewer as V } from "./annotorious-react.es8.js";
import { inline as j, offset as z, flip as H, shift as T, arrow as X } from "./annotorious-react.es16.js";
import { autoUpdate as Y } from "./annotorious-react.es15.js";
import w from "./annotorious-react.es17.js";
const I = (n, r) => {
  const { minX: d, minY: a, maxX: t, maxY: m } = r.bounds, { top: s, left: p } = n.element.getBoundingClientRect(), o = n.viewport.imageToViewerElementCoordinates(new g.Point(d, a)), l = n.viewport.imageToViewerElementCoordinates(new g.Point(t, m));
  return new DOMRect(
    o.x + p,
    o.y + s,
    l.x - o.x,
    l.y - o.y
  );
}, re = (n) => {
  var u, f;
  const r = S(), [d, a] = P(!1), t = V(), m = A(null), { selected: s, event: p } = b(), o = (u = s[0]) == null ? void 0 : u.annotation, l = (f = s[0]) == null ? void 0 : f.editable, { refs: c, floatingStyles: y, context: v } = F({
    open: d,
    onOpenChange: a,
    placement: n.placement,
    middleware: [
      j(),
      z(10),
      H({ crossAxis: !0 }),
      T({
        crossAxis: !0,
        boundary: t == null ? void 0 : t.element,
        padding: { right: 5, left: 5, top: 10, bottom: 10 }
      }),
      X({
        element: m,
        padding: 5
      })
    ],
    whileElementsMounted: Y
  });
  D(() => {
    if (s.length === 0)
      a(!1);
    else {
      const e = () => {
        if (!t.element) return;
        const i = I(t, o.target.selector.geometry);
        c.setReference({
          getBoundingClientRect: () => i,
          getClientRects: () => U(i)
        });
      };
      return window.addEventListener("scroll", e, !0), window.addEventListener("resize", e), t.addHandler("update-viewport", e), e(), a(!0), () => {
        window.removeEventListener("scroll", e, !0), window.removeEventListener("resize", e), t.removeHandler("update-viewport", e);
      };
    }
  }, [n.popup, s, t]);
  const x = (e) => {
    const i = e.id || w();
    r.state.store.addBody({
      ...e,
      id: i,
      annotation: o.id,
      created: e.created || /* @__PURE__ */ new Date(),
      creator: r.getUser()
    });
  }, R = (e) => {
    r.state.store.deleteBody({ id: e, annotation: o.id });
  }, B = (e, i) => {
    const C = i.id || w(), E = {
      updated: /* @__PURE__ */ new Date(),
      updatedBy: r.getUser(),
      ...i,
      id: C,
      annotation: o.id
    };
    r.state.store.updateBody(e, E);
  };
  return d && o ? L(
    /* @__PURE__ */ h(
      "div",
      {
        className: "a9s-popup a9s-image-popup",
        ref: c.setFloating,
        style: y,
        children: [
          n.arrow && /* @__PURE__ */ O(
            M,
            {
              ref: m,
              context: v,
              ...n.arrowProps || {}
            }
          ),
          n.popup({
            annotation: o,
            editable: l,
            event: p,
            onCreateBody: x,
            onDeleteBody: R,
            onUpdateBody: B
          })
        ]
      }
    ),
    t.element
  ) : null;
};
export {
  re as OpenSeadragonAnnotationPopup
};
