// @annotorious/react/dist/annotorious-react.es2.js
import { jsx as j } from "react/jsx-runtime";
import { createContext as E, forwardRef as P, useState as l, useImperativeHandle as R, useEffect as h, useContext as w } from "react";
import { useDebounce as _ } from "./annotorious-react.es12.js";
const d = E({
  anno: void 0,
  setAnno: void 0,
  annotations: [],
  selection: { selected: [] }
}), K = P((t, e) => {
  const [n, c] = l(null), [a, o] = l([]), [i, r] = l({ selected: [] });
  return R(e, () => n), h(() => {
    if (n) {
      const { selection: S, store: s, hover: V } = n.state;
      s.all().length > 0 && o(s.all());
      const b = () => o(() => s.all());
      s.observe(b);
      let u;
      const A = S.subscribe(({ selected: f, ...I }) => {
        u && s.unobserve(u);
        const O = (f || []).map(({ id: v, editable: m }) => ({ annotation: s.getAnnotation(v), editable: m }));
        r({ selected: O, ...I }), u = (v) => {
          const { updated: m } = v.changes;
          r((g) => ({
            ...g,
            selected: g.selected.map(({ annotation: x, editable: C }) => {
              const H = m.find((U) => U.oldValue.id === x.id);
              return H ? { annotation: H.newValue, editable: C } : { annotation: x, editable: C };
            })
          }));
        }, s.observe(u, { annotations: f.map(({ id: v }) => v) });
      });
      return () => {
        s.unobserve(b), A();
      };
    }
  }, [n]), /* @__PURE__ */ j(d.Provider, { value: {
    anno: n,
    setAnno: c,
    annotations: a,
    selection: i
  }, children: t.children });
}), p = () => {
  const { anno: t } = w(d);
  return t;
}, k = () => {
  const t = p();
  return t == null ? void 0 : t.state.store;
}, q = () => {
  const { annotations: t } = w(d);
  return t;
}, y = (t) => {
  const { annotations: e } = w(d);
  return _(e, t);
}, L = (t) => t ? y(t) : q(), z = (t, e) => {
  const n = k(), [c, a] = l(
    n == null ? void 0 : n.getAnnotation(t)
  );
  return h(() => {
    if (!n) return;
    const o = (i) => {
      const r = i.changes.updated[0];
      r && a(r.newValue);
    };
    return n.observe(o, { ...e, annotations: t }), () => n.unobserve(o);
  }, []), c;
}, M = (t) => {
  const e = p(), n = z(t);
  return e && n ? e.state.selection.evalSelectAction(n) : void 0;
}, N = () => {
  const { selection: t } = w(d);
  return t;
}, Q = () => {
  const t = p(), [e, n] = l();
  return h(() => {
    if (!t) return;
    const { hover: c, store: a } = t.state, o = c.subscribe((i) => {
      if (i) {
        const r = a.getAnnotation(i);
        n(r);
      } else
        n(void 0);
    });
    return () => {
      o();
    };
  }, [t]), e;
}, T = () => {
  const t = p();
  return t == null ? void 0 : t.getUser();
}, D = () => {
  const t = p(), [e, n] = l([]);
  return h(() => {
    if (t) {
      const { store: c, viewport: a } = t.state;
      if (!a)
        return;
      let o;
      const i = a.subscribe((r) => {
        o && c.unobserve(o);
        const S = r.map((s) => c.getAnnotation(s));
        n(S), o = (s) => {
          const { updated: V } = s.changes;
          n((b) => b.map((u) => {
            const A = V.find((f) => f.oldValue.id === u.id);
            return A ? A.newValue : u;
          }));
        }, c.observe(o, { annotations: r });
      });
      return () => {
        i();
      };
    }
  }, [t]), e;
}, B = (t) => {
  const e = D();
  return _(e, t);
}, W = (t) => t ? B(t) : D();
export {
  K as Annotorious,
  d as AnnotoriousContext,
  z as useAnnotation,
  M as useAnnotationSelectAction,
  k as useAnnotationStore,
  L as useAnnotations,
  p as useAnnotator,
  T as useAnnotatorUser,
  Q as useHover,
  N as useSelection,
  W as useViewportState
};
