// @recogito/react-text-annotator/dist/react-text-annotator.es.js
import { jsx as u, Fragment as L, jsxs as N } from "react/jsx-runtime";
import { useRef as p, useEffect as a, useContext as S, useCallback as M, Children as k, cloneElement as D, useState as v, useMemo as Q } from "react";
import W from "CETEIcean";
import { AnnotoriousContext as O, useAnnotationStore as _, Origin as j, useAnnotator as H, useSelection as U } from "@annotorious/react";
import { AnnotoriousPlugin as Re } from "@annotorious/react";
import { TEIPlugin as V } from "@recogito/text-annotator-tei";
import { serializeW3CTEIAnnotation as Fe } from "@recogito/text-annotator-tei";
import { createTextAnnotator as P, debounce as z, toDomRectList as Y, NOT_ANNOTATABLE_CLASS as $ } from "@recogito/text-annotator";
import { W3CTextFormat as Ie, parseW3CTextAnnotation as Le, serializeW3CTextAnnotation as Ne } from "@recogito/text-annotator";
import "@recogito/text-annotator/dist/text-annotator.css";
import { useFloating as q, autoUpdate as J, inline as G, offset as K, flip as X, shift as Z, arrow as ee, useDismiss as te, useRole as ne, useInteractions as oe, FloatingPortal as re, FloatingFocusManager as se, FloatingArrow as ie } from "@floating-ui/react";
import { Ignore as le } from "@annotorious/core";
import { Origin as ke, UserSelectAction as De, createBody as Qe } from "@annotorious/core";
const x = {
  tei: {
    ref: (e) => {
      const t = document.createElement("a");
      for (; e.firstChild; )
        t.appendChild(e.removeChild(e.firstChild));
      t.setAttribute("href", e.getAttribute("target")), e.appendChild(t);
    },
    // Cf. https://github.com/TEIC/CETEIcean/issues/67
    graphic: (e) => {
      var o;
      const t = new Image();
      t.src = (o = e.getAttribute("url")) == null ? void 0 : o.trim(), e.hasAttribute("width") && t.setAttribute("width", e.getAttribute("width")), e.hasAttribute("height") && t.setAttribute("height", e.getAttribute("height")), e.appendChild(t);
    },
    list: null,
    note: null,
    table: null,
    teiHeader: (e) => {
      e.hidden = !0;
    }
  }
}, xe = (e) => {
  const t = p(null);
  return a(() => {
    var o;
    if (e.tei) {
      const r = new W(e.initArgs);
      r.addBehaviors({
        ...x,
        ...e.behaviors || {},
        tei: {
          ...x.tei,
          ...((o = e.behaviors) == null ? void 0 : o.tei) || {}
        }
      }), r.makeHTML5(e.tei, (n) => {
        t.current.appendChild(n), e.onLoad(t.current);
      });
    }
    return () => {
      t.current && (t.current.innerHTML = "");
    };
  }, [e.tei, JSON.stringify(e.initArgs), e.onLoad]), /* @__PURE__ */ u(
    "div",
    {
      ref: t,
      className: "tei-container"
    }
  );
}, ye = (e) => {
  const { children: t, ...o } = e, { anno: r, setAnno: n } = S(O), l = M((s) => {
    const c = V(P(s, o));
    n(c);
  }, []);
  return a(() => {
    r && r.setStyle(e.style);
  }, [e.style]), a(() => {
    r && r.setFilter(e.filter);
  }, [e.filter]), e.children ? /* @__PURE__ */ u(L, { children: k.toArray(e.children).map((s) => D(s, { onLoad: l })) }) : null;
}, Se = (e) => {
  const t = p(null), { className: o, children: r, ...n } = e, { style: l, filter: s, user: c } = n, { anno: i, setAnno: d } = S(O);
  return a(() => {
    if (!d) return;
    const f = typeof n.adapter == "function" ? n.adapter(t.current) : n.adapter, g = P(t.current, { ...n, adapter: f });
    return d(g), () => g.destroy();
  }, [d]), a(() => i == null ? void 0 : i.setStyle(l), [i, l]), a(() => i == null ? void 0 : i.setFilter(s), [i, s]), a(() => i == null ? void 0 : i.setUser(c), [i, c]), /* @__PURE__ */ u("div", { ref: t, className: `r6o-annotatable no-focus-outline ${o}`, children: r });
}, y = () => {
  if (typeof navigator > "u") return !1;
  var e = navigator.userAgent || navigator.vendor || (window == null ? void 0 : window.opera);
  return !!(/android/i.test(e) || /iPad|iPhone/.test(e) && !window.MSStream);
}, ae = (e, t = { timeout: 250 }) => {
  const o = _(), [r, n] = v(!0);
  return a(() => {
    if (!o || !e) return;
    let l;
    const s = (c) => {
      const { changes: { updated: i } } = c;
      (i == null ? void 0 : i.some((f) => {
        const { targetUpdated: g } = f;
        if (g) {
          const { oldTarget: w, newTarget: m } = g;
          return ce(w, m);
        }
      })) && (n(!1), clearTimeout(l), l = setTimeout(() => n(!0), t.timeout));
    };
    return o.observe(s, {
      annotations: e,
      ignore: le.BODY_ONLY,
      origin: j.LOCAL
    }), () => {
      clearTimeout(l), o.unobserve(s);
    };
  }, [o, e, t.timeout]), r;
}, ce = (e, t) => {
  const { selector: o } = e, r = o.map(({ quote: s }) => s), { selector: n } = t, l = n.map(({ quote: s }) => s);
  return r.join() !== l.join();
};
let C = null;
const R = (e, t) => {
  const { left: o, top: r, right: n, bottom: l } = e, s = t.getBoundingClientRect();
  return new DOMRect(o + s.left, r + s.top, n - o, l - r);
}, ue = z((e, t, o) => {
  requestAnimationFrame(() => {
    const r = t.getAnnotationBounds(e);
    r && (C = R(r, o));
  });
}, 250), de = (e) => {
  var b;
  const t = H(), { selected: o, event: r } = U(), n = (b = o[0]) == null ? void 0 : b.annotation, l = ae(n == null ? void 0 : n.id), [s, c] = v((o == null ? void 0 : o.length) > 0), i = p(!1), d = p(null), { refs: f, floatingStyles: g, update: w, context: m } = q({
    placement: y() ? "bottom" : e.placement || "top",
    strategy: "absolute",
    open: s,
    onOpenChange: (h, T, A) => {
      !h && (A === "escape-key" || A === "focus-out") && (c(h), t == null || t.cancelSelected());
    },
    middleware: [
      G(),
      K(10),
      X({ crossAxis: !0 }),
      Z({ crossAxis: !0, padding: 10 }),
      ee({ element: d })
    ],
    whileElementsMounted: J
  }), E = te(m), F = ne(m, { role: "dialog" }), { getFloatingProps: fe } = oe([E, F]);
  a(() => {
    if (n != null && n.id && l) {
      const h = t == null ? void 0 : t.state.store.getAnnotationBounds(n.id);
      c(!!h);
    } else
      c(!1);
  }, [n == null ? void 0 : n.id, n == null ? void 0 : n.target.selector, l, t == null ? void 0 : t.state.store]), a(() => {
    e.onClose && (s ? i.current = !0 : i.current && (i.current = !1, e.onClose()));
  }, [e.onClose, s]), a(() => {
    t && (s && (n != null && n.id) ? f.setPositionReference({
      getBoundingClientRect: () => (ue(n.id, t.state.store, t.element), C || new DOMRect()),
      getClientRects: () => {
        const T = t.state.store.getAnnotationRects(n.id).map((A) => R(A, t.element));
        return Y(T);
      }
    }) : f.setPositionReference(null));
  }, [s, n == null ? void 0 : n.id, n == null ? void 0 : n.target, t]);
  const B = Q(() => (r == null ? void 0 : r.type) === "keyup" || (r == null ? void 0 : r.type) === "contextmenu" || y() ? -1 : 0, [r]), I = () => t == null ? void 0 : t.cancelSelected();
  return s && n ? /* @__PURE__ */ u(
    re,
    {
      root: e.asPortal ? null : t.element,
      children: /* @__PURE__ */ u(
        se,
        {
          context: m,
          modal: !1,
          closeOnFocusOut: !0,
          returnFocus: !1,
          initialFocus: B,
          children: /* @__PURE__ */ N(
            "div",
            {
              className: `a9s-popup r6o-popup annotation-popup r6o-text-popup ${$}`,
              ref: f.setFloating,
              style: g,
              children: [
                e.popup({
                  annotation: o[0].annotation,
                  editable: o[0].editable,
                  event: r
                }),
                e.arrow && /* @__PURE__ */ u(
                  ie,
                  {
                    ref: d,
                    context: m,
                    ...e.arrowProps || {}
                  }
                ),
                /* @__PURE__ */ u("button", { className: "r6o-popup-sr-only", "aria-live": "assertive", onClick: I, children: e.ariaCloseWarning || "Click or leave this dialog to close it." })
              ]
            }
          )
        }
      )
    }
  ) : null;
}, ve = (e) => (a(() => {
  console.warn("TextAnnotatorPopup is deprecated and will be removed in a future version. Please use TextAnnotationPopup instead.");
}, []), /* @__PURE__ */ u(de, { ...e }));
export {
  Re as AnnotoriousPlugin,
  xe as CETEIcean,
  ke as Origin,
  ye as TEIAnnotator,
  de as TextAnnotationPopup,
  Se as TextAnnotator,
  ve as TextAnnotatorPopup,
  De as UserSelectAction,
  Ie as W3CTextFormat,
  Qe as createBody,
  Le as parseW3CTextAnnotation,
  Fe as serializeW3CTEIAnnotation,
  Ne as serializeW3CTextAnnotation
};
